Ensure that Gc.minor_words remains accurate after a GC.

stedolan opened the pull request:
If an allocation fails, the decrement of young_ptr should be undone before the GC is entered. This happened correctly on bytecode but not on native code. This commit fixes it for amd64 native.

~~This PR shouldn't be merged yet - it's probably broken on other native platforms as well. This patch adds a regression test, so we can see which platforms don't count minor_words correctly by seeing which builds are broken by this patch.~~ I think this is ready to merge now!

Closes #7798.

stedolan committed:
Ensure that Gc.minor_words remains accurate after a GC.

If an allocation fails, the decrement of young_ptr should be
undone before the GC is entered. This happened correctly on bytecode
but not on native code. This commit fixes it for amd64 native.

This is a partial cherry-pick of commit 8ceec on multicore.

dra27 commented:
Running through precheck - [job #246](https://ci.inria.fr/ocaml/job/precheck/246/)

stedolan committed:
Ensure that Gc.minor_words remains accurate after a GC (i386).

xavierleroy commented:
Precheck crashed, but I restarted it.

xavierleroy commented on asmcomp/amd64/emit.mlp:
```diff
@@ -290,6 +291,7 @@ let emit_call_gc gc =
     assert Config.spacetime;
     spacetime_before_uninstrumented_call ~node_ptr ~index
   end;
+  I.add (int gc.gc_size) r15;
```
Even though this `add` is out-of-line and in a cold part of the code, you're increasing the overall code size and making the code cache less efficient.  This will have to be benchmarked.

xavierleroy reviewed (commented):
I'm concerned that this can have a negative impact on performance, see below why.  Is it THAT important that `Gc.minor_words` is accurate?

xavierleroy commented:
The test fails (as could be expected) on ARM, ARM64, PowerPC, and System Z.

stedolan commented:
I think yes, for some applications it is important to be able to accurately measure allocations. It's certainly important for the multicore GC, which needs a similar patch (without this patch, there are holes of uninitialised data in the minor heap, which confuse multicore's promotion).

I don't think the code size increase is significant. I compared the code size before and after on the compiler distribution ([full data here](https://gist.github.com/stedolan/1223026073fd08bf6b287ee1312c6c97)). In total, `.text` sections got 1.7% bigger, with the biggest outlier being `asmcomp/x86_dsl.o`, whose `.text` section got 4% bigger.

stedolan commented:
> The test fails (as could be expected) on ARM, ARM64, PowerPC, and System Z.

Thanks! Happy to fix those, but I'll wait until there's consensus on whether this should be merged.

mshinwell commented:
I would add that when we investigated this behaviour at Jane Street a couple of years ago, it was in response to users being confused about why the minor words measurement wasn't accurate -- so this was user-driven rather than just spotting a technicality.

xavierleroy commented:
Thanks for the explanations and for the measurements.  I buy the "necessary for Multicore OCaml" argument.  

The increase in code size is less than I feared.  Indeed, for ocamlopt.opt the whole TEXT section increases by less than 1%.

It is possible to reduce the increase to negligible (0.1% overall, 0 for `asmcomp/x86_dsl.o`) by adding and using a few alternate entry points for `caml_call_gc`:

In `emit_call_gc`:
```
  begin match gc.gc_size with
  | 16 -> emit_call "caml_call_gc_1"
  | 24 -> emit_call "caml_call_gc_2"
  | 32 -> emit_call "caml_call_gc_3"
  | 40 -> emit_call "caml_call_gc_4"
  | 48 -> emit_call "caml_call_gc_5"
  | _  -> I.add (int gc.gc_size) r15;
          emit_call "caml_call_gc"
  end;
```

In runtime/amd64.S:
```
FUNCTION(G(caml_call_gc_1))
CFI_STARTPROC
        addq    $16, %r15
        jmp     G(caml_call_gc)
CFI_ENDPROC
```
(and likewise for the other 4 special entry points).

5 entry points is probably overkill.  But as you noticed I'm a stickler for code size.

gasche commented:
(I think that @jhjourdan probably looked at this question when he implemented statmemprof, so I'm pinging him in case the changes here imply some required changes for his code -- possibly simplifications.)

stedolan committed:
Add caml_call_gc{1,2,3} for code size optimisation

stedolan committed:
update Changes

stedolan commented:
@xavierleroy 
> It is possible to reduce the increase to negligible (0.1% overall, 0 for asmcomp/x86_dsl.o) by adding and using a few alternate entry points for caml_call_gc:

Good idea, thanks!

> 5 entry points is probably overkill.

I've added entry points for 1/2/3-word allocations, since those are the sizes special-cased when optimising for size with `-compact`. (If changing 4/5-word allocations makes a significant difference, we should probably change `-compact`).

@gasche 
This is no coincidence! I've started looking at the statmemprof code to try and get it merged (and am in contact with @jhjourdan). I don't think this particular PR should affect statmemprof, though.

stedolan committed:
typo in amd64.S

xavierleroy commented:
I had a very quick look at the other architectures.  ARM, ARM64 and System Z can be fixed exactly like x86-64.  PowerPC needs more work.  I could look into it or we could decide that PowerPC is not worth the effort.

stedolan commented:
I'd like to make this bug go away entirely, if possible. I'll have a go at fixing the other archs early next week. Is there something especially hairy about allocations on PowerPC?

xavierleroy commented:
> Is there something especially hairy about allocations on PowerPC?

Just look at `asmcomp/powerpc/emit.mlp`.  The glue code that calls the GC
- comes in 3 different versions (ppc32, ppc64, ppc64le)
- is much bigger than usual (for ppc64 and ppc64le)
- but is shared between all allocation sites, owing to the cool "conditional function call" instruction of PowerPC.

I can see plausible ways to fix the allocation pointer.  But there's the extra issue that our ppc32/ppc64 test machine (a PowerMac G5) died recently, leaving only ppc64le as testable.

mshinwell commented:
I think I have access to a system that can do ppc64 and possibly ppc32 as well.

xavierleroy committed:
Fix the Power/PowerPC port

Introduce one GC call point per allocation size.
Each call point corrects the allocation pointer r31 before calling caml_call_gc.
Tested on ppc64le (little-endian, ELF64v2 ABI).

Alternate code generation for calling the GC

xavierleroy commented:
I pushed a fix for PPC directly on this branch.  The generated code is quite compact, I'm pleased :-)  It was tested on ppc64le, but the changes are not specific to any of the 3 ppc variants, so with luck it should work for ppc32 and ppc64 as well.

gasche commented on asmcomp/power/emit.mlp:
```diff
@@ -1041,14 +1062,24 @@ let fundecl fundecl =
+           (sz1 - sz2, lbl1) :: delta_encode l
+      | ([] | [(_,_)]) as l -> l in
+    let delta_lbl_list =
+      delta_encode (List.rev (IntMap.bindings !call_gc_labels)) in
```
This encoding technique uses `n` instructions for declarations of `n` different sizes (one addition per label), instead of `2n+1` (one addition and a jump), at the cost of making each call slide through up to `n` different additions, The innocuous `List.rev` call here will have a noticeable impact on performance, as it guarantees that the largest calls are at the beginning of the addition slider, taking more time, and the (much more common) shortest calls are at the end, taking less time.

I think it could be nice to comment on the explicit choice of order here.

(Would it make sense to handle the most common allocations, sizes 1-5, with an extra jump, and use the slider approach only for larger sizes? (Separating the two is easy with `IntMap.split 6`, but of course this requires to implement the two logics.). I suppose that you didn't do this because there are fewer users of the architecture, so "simple" is better than "optimized".)

xavierleroy replied:
Generally speaking, I have the impression (not yet backed by numbers) that the number N of different allocation sizes per function is small, even when the function is quite large, hence the cost of going through at most N additions is very small.  Also, remember that few allocation points call the GC, signal handling notwithstanding.

The main reason for sorting allocation sizes in decreasing order was to make the generated code more legible: all deltas are positive, hence we can "see" the allocation pointer being increased until it reaches its initial value.  Any other order would be semantically correct but look odd, with the allocation pointer being decreased sometimes, increased some other times.  

The fact that smaller allocations go through a shorter sequence of adds was not planned, but is a nice touch :-)  Thank you for pointing it out.

I promise I'll put a comment.

xavierleroy replied:
The comment was added.

gasche commented on asmcomp/power/emit.mlp:
```diff
@@ -393,8 +392,38 @@ let name_for_specific = function
+      let lbl = 
+        match label_after_call_gc with Some l -> l | None -> new_label() in
+      call_gc_labels := IntMap.add sz lbl !call_gc_labels;
+      lbl
```
Is it really the correct logic to share the `label_after_call_gc` labels for all subsequent calls of the same size, and to not-place these labels if another label already exists in the table for this size? This explicit labelling choice comes from `asmcomp/spacetime_profiling` and I'm not sure what it does, but it looks the label is used to locate the address of an instrumented call in some runtime profiling structure, which may get confused by the sharing/erasing happening here?

xavierleroy replied:
I don't claim I fully understand the "label_after_call_gc" business and the way Spacetime profiling relies on it.  The code above is patterned after the what the PowerPC code generator currently does, namely:

* The (unique) GC call point is labeled with the "label_after_call_gc" carried by the first Ialloc instruction emitted, if any; otherwise, a fresh label is used.

I don't know whether that's correct w.r.t. Spacetime --- would love to hear @mshinwell about this --- but it's what the current code does.  My code follows a very similar logic:

* The GC call point for size N is labeled with the "label_after_call_gc" carried by the first Ialloc(N) instruction emitted, if any; otherwise, a fresh label is used.

For this reason, I hope it's as good (or no worse) than the current code.  But that's for a Spacetime expert to confirm.

gasche replied:
On other backends (I only looked at amd64) `label_after_call_gc` is used in the `record_frame` function, which if I understand correctly emits call-specific metadata. Looking at the current PowerPC code more after your comment, my conclusion is that the handling of `label_after_call_gc` is probably already broken in its current state (different allocation points can use different labels, the latest to set the global ` call_gc_label` difference will win and the other labels will never get declared). I guess that no one ever tried Spacetime on PowerPC.

xavierleroy replied:
I'll be happy to `assert (label_after_call_gc = None)` if it turns out that Spacetime was not intended to work on PowerPC.

xavierleroy replied:
The mystery of Spacetime on PowerPC remains, but can be solved later.

stedolan committed:
Ensure that Gc.minor_words remains accurate after a GC (arm64)

stedolan commented:
I've just pushed an ARM64 patch. I don't have an ARM64 machine handy to test on right now, so I'd like to see whether this passes inria CI.

I'm using another approach: instead of mutating and then fixing young_ptr, this version only mutates young_ptr on successful allocations, doing the intermediate calculations on the result register instead. I think this is simpler, and it causes no code size increase on ARM because the 3-address format means it's just as easy to subtract from young_ptr and store the result elsewhere as to mutate young_ptr.

stedolan commented:
Incidentally, the arm64 allocation code contains a special sequence to handle allocations of more than 0xfff bytes. The largest possible allocation is 0x808 bytes, so this can never run. Anyone mind if I delete this code path? Or there a possibility of Max_young_wosize increasing significantly?

xavierleroy commented:
CI precheck in progress.

Concerning allocations of more than 0xffff bytes: I guess I didn't realize Max_young_wosize is much lower than 0xffff...  I'm OK with removing the large allocation case and replace it with an assert.

xavierleroy commented:
CI precheck is successful for arm64 and for ppc64le.

gasche commented:
@xavierleroy: I don't know if you missed it, or you are secretly working on it, or you checked that there is no problem, but [my comment above](https://github.com/ocaml/ocaml/pull/8619#discussion_r277278919) points at a potential bug in your PowerPC patch when Spacetime is enabled.

xavierleroy commented:
> a potential bug in your PowerPC patch when Spacetime is enabled.

See my explanations above.  The whole discussion is probably hypothetical, as the configure script limits Spacetime support to one architecture, x86 64 bits.

stedolan committed:
Ensure that the test for #7798 also tests -compact mode.

stedolan committed:
Ensure that Gc.minor_words remains accurate after a GC (arm, arm64, s390x)

stedolan commented:
I've just pushed changes for arm, arm64 and s390x. (all untested), based on the approach in my previous arm64 patch (subtract Wosize from young_ptr and check <= limit). None of these patches increase code size.

There was a bug in my previous arm64 patch: I'd only updated the fastcode path, not the `-compact` path. The regression test now explicitly tests `-compact` mode as well, so I'd like to rerun the tests even on architectures that previously passed.

stedolan committed:
fix typo in arm/emit.mlp (variable shadowing)

xavierleroy commented:
CI precheck in progress.

xavierleroy committed:
Fix s390x code generator

`lay` must be used instead of `la` for negative immediates.
Branch if less than or equal is brcl 12.

xavierleroy committed:
Fix ARM code generator

Trivial typo.

xavierleroy commented:
arm was broken because of a trivial typo, which I fixed in 745c2ef.
s390x was broken because of wrong asm code, which I fixed in 5d61fdd.
i386 (x86 32 bits) seems broken, but I haven't investigated yet.
amd64 (x86 64 bits) is fine.
arm64 is fine.

xavierleroy commented:
And ppc64le is fine.

stedolan commented:
Thanks! I suspect that arm typo was caused by emacs getting a ctrl-T that was intended for chrome. I'll figure out the i386 one, I can at least run that locally.

For future reference, where should I look up s390x instructions? I got `brcl 13` from [here](https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.2.0/com.ibm.zos.v2r2.asma400/asmr1021108.htm).

xavierleroy commented:
> For future reference, where should I look up s390x instructions? I got brcl 13 from here.

The full story is the Principles of Operation manual referenced here https://github.com/ocaml/ocaml/blob/trunk/asmcomp/s390x/NOTES.md .  But it's a big manual.

I got the "12" by looking at how integer comparisons are compiled already, especially the `branch_for_comparison` function https://github.com/ocaml/ocaml/blob/f845abefc92cec610e987698500bb56a6b6a261d/asmcomp/s390x/emit.mlp#L241
For a "less than" comparison it gives code 4, and for a "less than or equal" comparison it gives code 12.  

The one-line comment suggests what's going on: the branch conditional instruction uses the magic number as a mask against the (eq, lt, gt, ov) flags, branching if any of the flags with a 1 in the mask is set.  So, 12 means "branch if eq or lt", and 13 means "branch if eq or lt or overflow".  The "overflow" flag is important for FP comparisons, as it stands for "unordered", hence 13 means "branch if not greater than".  For integer comparisons, it's not obvious to me whether the "overflow" flag is cleared or left unchanged, so I'd rather not test it.

jhjourdan commented:
> @gasche
> This is no coincidence! I've started looking at the statmemprof code to try and get it merged (and am in contact with @jhjourdan). I don't think this particular PR should affect statmemprof, though.

I think it will, because statmemprof does a precise accounting of the allocation pointer in the minor heap in order to accurately sample memory blocks according to the desired probability distribution. I am certain that the corresponding change will be minimal, though.

stedolan committed:
Fix i386 -compact camlallocN code

stedolan committed:
whitespace

dra27 committed:
Fix MSVC64

dra27 committed:
Fix MSVC32

dra27 commented:
MSVC should now be fixed - it's running through [precheck](https://ci.inria.fr/ocaml/job/precheck/255/)

dra27 commented:
ARM is still failing:

```
List of failed tests:
    tests/regression/pr7798/'pr7798.ml' with 2 (native) 
    tests/lib-obj/'with_tag.ml' with 1 (native)
```

xavierleroy committed:
Fix ARM code generator

In the fast path for Ialloc, 4 extra bytes were allocated.

xavierleroy commented:
I fixed the ARM code generator.  4 too many bytes were allocated in the fast path of Ialloc, turning the minor heap into Swiss cheese.

stedolan commented:
Thanks for the fix!

stedolan committed:
Changes

xavierleroy commented:
Running through precheck again.

xavierleroy committed:
Add comment to explain the code generation strategy for calling the GC

As promised during review.

xavierleroy committed:
Simplify assembly_code_for_allocation

As noticed by @stedolan, the maximal size for an allocation in the
minor heap is well under 0x1000, hence the generated code need not
make provisions for bigger allocations.

xavierleroy commented:
Precheck seems happy.  However, I pushed two additional commits, as discussed during review,.  90f80dd adds a comment explaining the PowerPC code generation strategy. c6e64b2 simplifies the allocation sequence on ARM64, based on the fact that minor heap allocations are always < 0x10000 bytes in size.  So, I'll run precheck again.

xavierleroy committed:
Update Changes for 8619

xavierleroy reviewed (approved):
CI is happy, and I read the whole diff once again.  The changes are too large already for me to swear they are correct, but this is the best we can do, so let's merge.

gasche commented:
(Release pseudo-management on report: this is arguably a bugfix but it is large and touchy, so we are not putting this in 4.09.)
